### PR TITLE
Fixing parser error with read card

### LINKED ISSUE #206: SLY Syntax Error for Numbers in 'READ' cards
In GitLab by @tjlaboss on Jan 10, 2024, 10:25

MCNP allows numbers 0123456789 in `READ` cards, but this causes a syntax error when parsing in MontePy.

#### MWE
 _/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/_

_testRead.imcnp_:
```
Testing read card
C this is a comment
read file=file2read.imcnp

1 so 0.5 

mode n
```

_file2read.imcnp_:
```
1 0 -1
c
```

Traceback:

```python
Traceback (most recent call last):
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/TestRead.py", line 3, in <module>
    deck = montepy.read_input(fname)
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/input_parser/input_reader.py", line 26, in read_input
    problem.parse_input()
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/mcnp_problem.py", line 284, in parse_input
    raise e
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/mcnp_problem.py", line 269, in parse_input
    obj = obj_parser(input)
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/cell.py", line 82, in __init__
    super().__init__(input, self._parser)
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/mcnp_object.py", line 57, in __init__
    raise ParsingError(
montepy.errors.ParsingError:     testRead.imcnp, line 3

         2| C this is a comment
    >    3| read file=file2read.imcnp
          |  ^^^^ not expected here.
There was an error parsing "read".
sly: Syntax error at line 2, token=KEYWORD

```

### PR DESCRIPTION
**Pull request by MicahGale**
Problem 
-------

To recap the issue. The file:

```
Testing read card
C this is a comment
read file=file2read.imcnp

1 so 0.5 

mode n
```

Causes this error: 

``` python
Traceback (most recent call last):
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/TestRead.py", line 3, in <module>
    deck = montepy.read_input(fname)
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/input_parser/input_reader.py", line 26, in read_input
    problem.parse_input()
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/mcnp_problem.py", line 284, in parse_input
    raise e
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/mcnp_problem.py", line 269, in parse_input
    obj = obj_parser(input)
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/cell.py", line 82, in __init__
    super().__init__(input, self._parser)
  File "/projects/experiment_analysis/tjlaboss/mcnpy_testing/numberead/montepy/mcnp_object.py", line 57, in __init__
    raise ParsingError(
montepy.errors.ParsingError:     testRead.imcnp, line 3

         2| C this is a comment
    >    3| read file=file2read.imcnp
          |  ^^^^ not expected here.
There was an error parsing "read".
sly: Syntax error at line 2, token=KEYWORD
```

However after some looking into this; it is a red hearing. This is being thrown by the `CellParser`, and not the `ReadParser`. But why is that?

The key logic to decide if an input is a `ReadInput` is here in `mcnp_input.py`:

``` python
        parse_result = self._parser.parse(self.tokenize(), self)
        first_word = input_lines[0].split()[0].lower()
        if not parse_result:
            if first_word != "read":
                raise ValueError("Not a valid Read Input")
            else:
                raise ParsingError(self, "", self._parser.log.clear_queue())
```

So this is really a parsing error with `ReadInput` that is hidden. 

The real error is:

``` python 
>               raise ParsingError(self, "", self._parser.log.clear_queue())
E               montepy.errors.ParsingError:     tests/inputs/readEdgeCase.imcnp, line 3
E
E                        2| C this is a comment
E                   >    3| read file=file2read.imcnp
E                         |                ^^^^^ not expected here.
E               There was an error parsing "2read".
E               sly: Syntax error at line 2, token=NUMBER_WORD

montepy/input_parser/mcnp_input.py:304: ParsingError
```

Solution
--------

1. Don't use a parser to see if it a read input to increase efficiency.
2. Make `is_comment` a "global function"
3. Make searching for the keyword "read" more robust by discarding starting comments

Fixes #206.

### CODE REVIEW

**Comment by MicahGale** [comment]
Want to review this one first @tjlaboss?

**Comment by tjlaboss** [comment]
yes. I don't know when I'll get to it but I'll try to make some time in the immediate future

File: montepy/utilities.py [hunk 1]
```diff
@@ -31,6 +32,27 @@ def fortran_float(number_string):
...
+    :rtype: bool
+    """
+    upper_start = line[0 : BLANK_SPACE_CONTINUE + 1].upper()
+    non_blank_comment = upper_start and line.lstrip().upper().startswith("C ")
+    if non_blank_comment:
+        return True
+    blank_comment = (
+        "C\n" == upper_start.lstrip()
+        or "C\r\n" == upper_start.lstrip()
+        or ("C" == upper_start and "\n" not in line)
```
Comment (tjlaboss) [inline montepy/utilities.py:51]: Why not just `"C".strip()`?

Comment (MicahGale) [reply montepy/utilities.py:51]: I'd have to dig through the git history, but there were some edge cases that didn't work for this IIRC. It was really annoying and this isn't clean, but it works for all edge cases so far.

Comment (tjlaboss) [reply montepy/utilities.py:51]: If you can find one of the edge cases that breaks `"C".strip()`, think about discussing that in the function docstring or a code comment.

File: montepy/utilities.py [hunk 2]
```diff
@@ -31,6 +32,27 @@ def fortran_float(number_string):
...
+    )
+    return blank_comment or non_blank_comment
```
Comment (tjlaboss) [inline montepy/utilities.py:53]: Can just `return blank_comment` because of lines 46-47.

**Review by tjlaboss — CHANGES_REQUESTED** [review]

File: montepy/utilities.py (hunk 1 above)
Comment (MicahGale) [reply montepy/utilities.py]: Ok I think you were right. Is this new diff what you were thinking?

Comment (tjlaboss) [reply montepy/utilities.py]: Yeah that's better

**Review by tjlaboss — APPROVED** [review]

**Pull request merged**